Apps/W1/Shopify/app/src/Products/Codeunits/ShpfySalesChannelAPI.Codeunit.al

```diff
@@ -0,0 +1,95 @@
+        JResponse := CommunicationMgt.ExecuteGraphQL(GraphQLType::GetSalesChannels);
+        if JsonHelper.GetJsonArray(JResponse, JPublications, 'data.publications.edges') then
+            ProcessPublications(JPublications, ShopCode);
+    end;
+
+    /// <summary>
+    /// Processes the sales channels from Shopify.
+    /// </summary>
+    /// <param name="JPublications">The array of sales channels data from Shopify.</param>
+    /// <param name="ShopCode">The code of the shop.</param>
+    internal procedure ProcessPublications(JPublications: JsonArray; ShopCode: Code[20])
+    var
+        CurrentChannels: List of [BigInteger];
+    begin
+        CurrentChannels := CollectChannels(ShopCode);
+        InsertNewChannels(JPublications, ShopCode, CurrentChannels);
+        RemoveNotExistingChannels(CurrentChannels);
+    end;
+
+    local procedure CollectChannels(ShopCode: Code[20]): List of [BigInteger]
+    var
+        SalesChannel: Record "Shpfy Sales Channel";
+        Channels: List of [BigInteger];
+    begin
+        SalesChannel.SetRange("Shop Code", ShopCode);
+        if SalesChannel.FindSet() then
+            repeat
+                Channels.Add(SalesChannel.Id);
+            until SalesChannel.Next() = 0;
+        exit(Channels);
+    end;
+
+    local procedure RemoveNotExistingChannels(CurrentChannels: List of [BigInteger])
+    var
+        SalesChannel: Record "Shpfy Sales Channel";
+        ChannelId: BigInteger;
+    begin
+        if CurrentChannels.Count() <> 0 then
+            foreach ChannelId in CurrentChannels do begin
+                SalesChannel.Get(ChannelId);
+                SalesChannel.Delete(true);
+            end;
+    end;
+
+    local procedure InsertNewChannels(JPublications: JsonArray; ShopCode: Code[20]; CurrentChannels: List of [BigInteger])
+    var
+        SalesChannel: Record "Shpfy Sales Channel";
+        JPublication: JsonToken;
+        ChannelId: BigInteger;
+        JCatalogEdges: JsonArray;
+        JCatalogEdge: JsonToken;
+        Handle: Text;
+    begin
+        foreach JPublication in JPublications do begin
+            ChannelId := CommunicationMgt.GetIdOfGId(JsonHelper.GetValueAsText(JPublication, '$.node.id'));
+            if not SalesChannel.Get(ChannelId) then begin
+                SalesChannel.Init();
+                SalesChannel.Validate(Id, ChannelId);
+                JCatalogEdges := JsonHelper.GetJsonArray(JPublication, '$.node.catalog.apps.edges');
+                JCatalogEdges.Get(0, JCatalogEdge);
+                SalesChannel.Validate(Name, JsonHelper.GetValueAsText(JCatalogEdge, '$.node.title'));
+                Handle := JsonHelper.GetValueAsText(JCatalogEdge, '$.node.handle');
+                if Handle = 'online_store' then
+                    SalesChannel.Default := true;
+                SalesChannel."Shop Code" := ShopCode;
+                SalesChannel.Insert(true);
+            end else
+                CurrentChannels.Remove(ChannelId);
```

Comment:
shouldn't the CurrentChannels.Remove be called also in the case where SalesChannel gets created?

Otherwise, the Channel doesn't exist yet -> we create it -> we don't remove it from the CurrentChannels -> RemoveExisting deletes it becose the id is still in the list?

Reply:
Ahh, it's never in the CurrentChannels to begin with, because it doesn't exist... *facepalm*